src/Numerics/SystemSolvers/batched_generalized_minimal_residual_solver.jl

```diff
@@ -1,167 +1,188 @@
+        # Get ArrayType information
+        if isa(array_device(Q), CPU)
+            ArrayType = Array
+        else
+            # Sanity check since we don't support anything else
+            @assert isa(array_device(Q), CUDADevice)
+            ArrayType = CuArray
+        end
 
-# Keyword Arguments
-- `m`: (int) size of vector space for each independent linear solve. This is assumed to be the same for each and every linear solve. `DEFAULT = size(Qrhs)[1]`
-- `n`: (int) number of independent linear solves, `DEFAULT = size(Qrhs)[end]`
-- `atol`: (float) absolute tolerance. `DEFAULT = sqrt(eps(eltype(Qrhs)))`
-- `rtol`: (float) relative tolerance. `DEFAULT = sqrt(eps(eltype(Qrhs)))`
-- `ArrayType`: (type). used for either using CuArrays or Arrays. `DEFAULT = Array`
-- `reshape_tuple_f`: (tuple). used in the wrapper function for flexibility. `DEFAULT = size(Qrhs)`. this means don't do anything
-- `permute_tuple_f`: (tuple). used in the wrapper function for flexibility. `DEFAULT = Tuple(1:length(size(Qrhs)))`. this means, don't do anything.
-
-# Return
-instance of `BatchedGeneralizedMinimalResidual` struct
-"""
-function BatchedGeneralizedMinimalResidual(
-    Qrhs;
-    m = size(Qrhs)[1],
-    n = size(Qrhs)[end],
-    subspace_size = m,
-    atol = sqrt(eps(eltype(Qrhs))),
-    rtol = sqrt(eps(eltype(Qrhs))),
-    ArrayType = Array,
-    reshape_tuple_f = size(Qrhs),
-    permute_tuple_f = Tuple(1:length(size(Qrhs))),
-)
-    k_n = subspace_size
-    # define the back permutations and reshapes
-    permute_tuple_b = Tuple(sortperm([permute_tuple_f...]))
-    tmp_reshape_tuple_b = [reshape_tuple_f...]
-    permute!(tmp_reshape_tuple_b, [permute_tuple_f...])
-    reshape_tuple_b = Tuple(tmp_reshape_tuple_b)
-    # allocate memory
-    residual = ArrayType(zeros(eltype(Qrhs), (k_n, n)))
-    b = ArrayType(zeros(eltype(Qrhs), (m, n)))
-    x = ArrayType(zeros(eltype(Qrhs), (m, n)))
-    sol = ArrayType(zeros(eltype(Qrhs), (m, n)))
-    rhs = ArrayType(zeros(eltype(Qrhs), (k_n + 1, n)))
-    cs = ArrayType(zeros(eltype(Qrhs), (2 * k_n, n)))
-    Q = ArrayType(zeros(eltype(Qrhs), (m, k_n + 1, n)))
-    H = ArrayType(zeros(eltype(Qrhs), (k_n + 1, n)))
-    R = ArrayType(zeros(eltype(Qrhs), (k_n + 1, k_n, n)))
-    return BatchedGeneralizedMinimalResidual(
-        atol,
-        rtol,
-        m,
-        n,
-        k_n,
-        residual,
-        b,
-        x,
-        sol,
-        rhs,
-        cs,
-        Q,
-        H,
-        R,
-        reshape_tuple_f,
-        permute_tuple_f,
-        reshape_tuple_b,
-        permute_tuple_b,
-    )
+        # FIXME: If we can batch the application of linearoperator!, then we dont
+        # need these two temporary work vectors (unpermuted/reshaped)
+        krylov_basis = similar(Q)
+        krylov_basis_prev = similar(Q)
+
+        # Create storage for holding the batched Krylov basis
+        batched_krylov_basis =
+            ArrayType(zeros(eltype(AT), M + 1, dofperbatch, Nbatch))
```

Comment:
```suggestion
            fill!(ArrayType{FT}(undef, M + 1, dofperbatch, Nbatch), 0)
```

And elsewhere that this is done (no need to create an array and then a new copy of the array)

Reply:
Done!